scapy/contrib/isotp.py

```diff
@@ -0,0 +1,348 @@
+            raise Scapy_Exception(m)
+        return ifr
+
+    def __bindSocket(self, sock, iface, sid, did):
+        socketID = ctypes.c_int(sock.fileno())
+        ifr = self.__getSockIfreq(sock, iface)
+
+        if sid > 0x7ff:
+            sid = sid | socket.CAN_EFF_FLAG
+        if did > 0x7ff:
+            did = did | socket.CAN_EFF_FLAG
+
+        # select the CAN interface and bind the socket to it
+        addr = SOCKADDR_CAN(ctypes.c_uint16(socket.PF_CAN),
+                            ifr.ifr_ifindex,
+                            ADDR_INFO(TP(ctypes.c_uint32(sid),
+                                         ctypes.c_uint32(did))))
+
+        error = libc.bind(socketID, ctypes.byref(addr), ctypes.sizeof(addr))
+
+        if error < 0:
+            warning("Couldn't bind socket")
+
+    def __setOptionFlags(self, sock, extended_addr=None,
+                         extended_rx_addr=None,
+                         listen_only=False):
+        option_flags = CAN_ISOTP_DEFAULT_FLAGS
+        if extended_addr is not None:
+            option_flags = option_flags | CAN_ISOTP_EXTEND_ADDR
+        else:
+            extended_addr = CAN_ISOTP_DEFAULT_EXT_ADDRESS
+
+        if extended_rx_addr is not None:
+            option_flags = option_flags | CAN_ISOTP_RX_EXT_ADDR
+        else:
+            extended_rx_addr = CAN_ISOTP_DEFAULT_EXT_ADDRESS
+
+        if listen_only:
+            option_flags = option_flags | CAN_ISOTP_LISTEN_MODE
+
+        sock.setsockopt(SOL_CAN_ISOTP,
+                        CAN_ISOTP_OPTS,
+                        self.__build_can_isotp_options(
+                            flags=option_flags,
+                            ext_address=extended_addr,
+                            rx_ext_address=extended_rx_addr))
+
+    def __init__(self,
+                 iface=None,
+                 sid=0,
+                 did=0,
+                 extended_addr=None,
+                 extended_rx_addr=None,
+                 listen_only=False,
+                 basecls=ISOTP):
+        self.iface = conf.contribs['NativeCANSocket']['iface'] \
+            if iface is None else iface
+        self.ins = socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, CAN_ISOTP)
+        self.__setOptionFlags(self.ins,
+                              extended_addr,
+                              extended_rx_addr,
+                              listen_only)
+
+        self.ins.setsockopt(SOL_CAN_ISOTP,
+                            CAN_ISOTP_RECV_FC,
+                            self.__build_can_isotp_fc_options())
+        self.ins.setsockopt(SOL_CAN_ISOTP,
+                            CAN_ISOTP_LL_OPTS,
+                            self.__build_can_isotp_ll_options())
+
+        self.__bindSocket(self.ins, iface, sid, did)
+        self.outs = self.ins
+        if basecls is None:
+            warning('Provide a basecls ')
+        self.basecls = basecls
+
+    def recv(self, x=0xffff):
+        try:
+            pkt, sa_ll = self.ins.recvfrom(x)
```

Comment:
if you don't need `sa_ll`, you could rewrite this as `pkt = self.ins.recvfrom(x)[0]`